genai-function-calling: uses OpenInference to propagate trace IDs over mcp stdio

codefromthecrypt opened the pull request:
This adds [OpenInference instrumentation for MCP](https://github.com/Arize-ai/openinference/tree/main/js/packages/openinference-instrumentation-mcp), which connects the trace context between the MCP client and the server it launches as a sub-process.

Technically, this injects trace identifiers into the request.params._meta field on the client side, and extracts them on the server.

<img width="1679" alt="Screenshot 2025-04-25 at 11 47 39 AM" src="https://github.com/user-attachments/assets/59bb1877-a461-4938-a264-39f4b739d943" />

codefromthecrypt commented on genai-function-calling/vercel-ai/mcp.js:
```diff
@@ -1,16 +1,19 @@
 // Get MCP server parameters from package.json
 let name, version;
 try {
+    const __filename = fileURLToPath(import.meta.url);
```
ack I might be digging a hole on this point and might just switch it to hard-coded since the values don't matter in an example

codefromthecrypt commented:
to make this easier, I folded telemetry.js into main.js and turned `OTEL_LOG_LEVEL=debug` 

Then, I see a patch failed like this which fails like this "module TypeError: Cannot add property openInferencePatched, object is not extensible"

```
{"name":"elastic-otel-node","level":20,"msg":"Manually instrumenting @modelcontextprotocol/sdk/client/index","time":"2025-04-22T11:50:54.906Z"}
{"name":"elastic-otel-node","level":20,"msg":"Applying patch for @modelcontextprotocol/sdk/client/index@undefined","time":"2025-04-22T11:50:54.906Z"}
{"name":"elastic-otel-node","level":20,"msg":"Failed to set @modelcontextprotocol/sdk/client/index patched flag on the module TypeError: Cannot add property openInferencePatched, object is not extensible\n    at MCPInstrumentation._patchClientModule (file:///Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/node_modules/@arizeai/openinference-instrumentation-mcp/dist/esm/mcp.js:68:41)\n    at MCPInstrumentation.manuallyInstrument (file:///Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/node_modules/@arizeai/openinference-instrumentation-mcp/dist/esm/mcp.js:46:18)\n    at file:///Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/index.js:8:20\n    at ModuleJob.run (node:internal/modules/esm/module_job:271:25)\n    at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:578:26)\n    at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:116:5)","time":"2025-04-22T11:50:54.906Z"}
{"name":"elastic-otel-node","level":20,"msg":"Manually instrumenting @modelcontextprotocol/sdk/server/index","time":"2025-04-22T11:50:54.906Z"}
{"name":"elastic-otel-node","level":20,"msg":"Applying patch for @modelcontextprotocol/sdk/server/index@undefined","time":"2025-04-22T11:50:54.906Z"}
{"name":"elastic-otel-node","level":20,"msg":"Failed to set @modelcontextprotocol/sdk/server/index patched flag on the module TypeError: Cannot add property openInferencePatched, object is not extensible\n    at MCPInstrumentation._patchServerModule (file:///Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/node_modules/@arizeai/openinference-instrumentation-mcp/dist/esm/mcp.js:104:41)\n    at MCPInstrumentation.manuallyInstrument (file:///Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/node_modules/@arizeai/openinference-instrumentation-mcp/dist/esm/mcp.js:50:18)\n    at file:///Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/index.js:8:20\n    at ModuleJob.run (node:internal/modules/esm/module_job:271:25)\n    at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:578:26)\n    at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:116:5)","time":"2025-04-22T11:50:54.906Z"}
The latest version of Elasticsearch 8 is 8.18.0.
```

codefromthecrypt commented:
ok I changed to manually wire instead of using EDOT import and the MCP propagation works now. 

![Screenshot 2025-04-22 at 9 11 15 PM](https://github.com/user-attachments/assets/74173fb4-8534-4a04-8d25-be5d246720bb)

@trentm any ideas on this one? notably the comment I made [here](https://github.com/elastic/observability-examples/pull/64#issuecomment-2821092664) which latest commit works around?

"Cannot add property openInferencePatched, object is not extensible"

https://github.com/Arize-ai/openinference/tree/main/js/packages/openinference-instrumentation-mcp is the source of the experimental propagation instrumentation we're using to support an MCP spec discussion

anuraaga reviewed (commented):
Just for reference, the example here uses commonjs and seemed fine.

https://github.com/elastic/genai-instrumentation/blob/main/experiments%2Finstrumentation-mcp-js%2Fexamples%2Ftelemetry.js

But with node there are so many variables, can check things out at least by Thurs.

anuraaga commented:
The debug log is here

https://github.com/Arize-ai/openinference/blob/main/js%2Fpackages%2Fopeninference-instrumentation-mcp%2Fsrc%2Fmcp.ts#L188

All the openinference instrumentation includes it.

codefromthecrypt commented:
right cjs probably was a red herring, I'll try to revert esm now

codefromthecrypt commented:
ok I got it working with explicit SDK initialization with cjs (no split trace)

Then I switched back to EDOT, then I got this

```
{"name":"elastic-otel-node","level":40,"msg":"@arizeai/openinference-instrumentation-mcp Module @modelcontextprotocol/sdk/client/index.js has been loaded before @arizeai/openinference-instrumentation-mcp so it might not work, please initialize it before requiring @modelcontextprotocol/sdk/client/index.js","time":"2025-04-22T23:36:02.107Z"}
```

Then, I changed to this in telemetry.js (this code is in the most recent commit):

```javascript
// First, import and initialize the instrumentation
const { MCPInstrumentation } = require("@arizeai/openinference-instrumentation-mcp");
const mcpInstrumentation = new MCPInstrumentation();

// Then, import the MCP modules after instrumentation is set up
const MCPClientModule = require("@modelcontextprotocol/sdk/client/index.js");
const MCPServerModule = require("@modelcontextprotocol/sdk/server/index.js");

// Finally, manually instrument the modules
mcpInstrumentation.manuallyInstrument({
    clientModule: MCPClientModule,
    serverModule: MCPServerModule,
});
```

Now, I don't see any errors (at least on the client side), but the trace is split:
```
{"name":"elastic-otel-node","level":20,"msg":"Manually instrumenting @modelcontextprotocol/sdk/client/index","time":"2025-04-22T23:38:42.443Z"}
{"name":"elastic-otel-node","level":20,"msg":"Applying patch for @modelcontextprotocol/sdk/client/index@undefined","time":"2025-04-22T23:38:42.443Z"}
{"name":"elastic-otel-node","level":20,"msg":"Manually instrumenting @modelcontextprotocol/sdk/server/index","time":"2025-04-22T23:38:42.443Z"}
{"name":"elastic-otel-node","level":20,"msg":"Applying patch for @modelcontextprotocol/sdk/server/index@undefined","time":"2025-04-22T23:38:42.443Z"}
The latest version of Elasticsearch 8 is 8.18.0.
```

I'm not sure if this is something I am doing wrong or a limitation

trentm commented on genai-function-calling/vercel-ai/package.json:
```diff
@@ -7,13 +7,14 @@
   },
   "scripts": {
-    "start": "node --env-file .env -r @elastic/opentelemetry-node index.js"
+    "start": "node --env-file .env -r @elastic/opentelemetry-node -r ./telemetry.js index.js"
```
Not sure it matters now that you've gone back to CommonJS, but EDOT Node.js now recommends using
`node --import @elastic/opentelemetry-node ...` rather than `node -r @elastic/opentelemetry-node`. Only the former will enable instrumentation of ES modules.  Using `--import ...` is fine and safe to use even if you are using CommonJS for your code.

Anyway, I doubt this is related to the issues you are currently seeing.

codefromthecrypt replied:
fyi I rewrote this comment with a lengthy overview which might help https://github.com/elastic/observability-examples/pull/64#issuecomment-2822722800

codefromthecrypt replied:
thanks for the tip. I switched to --import and verify base case (`npm run start`) makes a trace connected (not using MCP), and as you guessed it doesn't impact `npm run start -- --mcp` which is disconnected still.

trentm commented:
Some behaviour is differ between:

1. `node --env-file .env -r @elastic/opentelemetry-node -r ./telemetry.js index.js`
2. `node --env-file .env --import @elastic/opentelemetry-node -r ./telemetry.js index.js`

In the second case (when using `--import` for EDOT), I don't see this log output:

```
{"name":"elastic-otel-node","level":20,"msg":"Manually instrumenting @modelcontextprotocol/sdk/client/index","time":"2025-04-23T00:11:53.671Z"}
{"name":"elastic-otel-node","level":20,"msg":"Applying patch for @modelcontextprotocol/sdk/client/index@undefined","time":"2025-04-23T00:11:53.671Z"}
{"name":"elastic-otel-node","level":20,"msg":"Manually instrumenting @modelcontextprotocol/sdk/server/index","time":"2025-04-23T00:11:53.671Z"}
{"name":"elastic-otel-node","level":20,"msg":"Applying patch for @modelcontextprotocol/sdk/server/index@undefined","time":"2025-04-23T00:11:53.671Z"}
```

A guess is that Node.js runs the `--require ...` before the `--import ...`, so the order has changed. And then the `mcpInstrumentation.manuallyInstrument(...)` in telemetry.js is `diag.debug(...)` logging before EDOT Node.js has installed a DiagLogger.

codefromthecrypt commented:
fyi this is a pending PR to fix the usage in openinference docs https://github.com/Arize-ai/openinference/pull/1551

trentm commented:
> Then, I changed to initialize `@arizeai/openinference-instrumentation-mcp` and there are no errors in the `OTEL_LOG_LEVEL=debug` output

FWIW, I think you can ignore those errors. I looked at the `manuallyInstrument` code and I think it will behave fine even if getting that debug message.

```ts
    try {
      // This can fail if the module is made immutable via the runtime or bundler
      module.openInferencePatched = true;
    } catch (e) {
      diag.debug(
        `Failed to set ${CLIENT_MODULE_NAME} patched flag on the module`,
        e,
      );
    }
```

I'd expect to get that debug message with ESM usage and perhaps with some bundlers.

trentm commented:
Possibly you could change "telemetry.js" to this:

```
require('@elastic/opentelemetry-node');

// Import and initialize MCPInstrumentation before importing the MCP modules.
const { MCPInstrumentation } = require("@arizeai/openinference-instrumentation-mcp");
const mcpInstrumentation = new MCPInstrumentation();

// Then, manually instrument MCP, as it doesn't have a traditional module structure.
mcpInstrumentation.manuallyInstrument({
    clientModule: require("@modelcontextprotocol/sdk/client/index.js"),
    serverModule: require("@modelcontextprotocol/sdk/server/index.js"),
});
```

Then you just need the one `-r ./telemetry.js` argument (instead of both `-r @elastic/opentelemetry-node -r ./telemetry.js`).

codefromthecrypt commented:
thanks for trying, but the last approach didn't join the trace :(

That said, I think I may have crossed wires at some point because I can't get an older commit to join the trace anymore (regardless of cjs  ESM). I should have caught a screen capture. I'll look at our internal demo and see if I can make a simpler reproduction.

codefromthecrypt commented:
ok this is the closest port I could make from the internal copy, and it isn't joining traces either... hmm

```javascript
// First, configure and start your OpenTelemetry NodeSDK per documentation.
const {NodeSDK} = require("@opentelemetry/sdk-node");
const {OTLPTraceExporter} = require("@opentelemetry/exporter-trace-otlp-proto");

const {diag, DiagConsoleLogger, DiagLogLevel} = require('@opentelemetry/api');
diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.DEBUG);

// Then, import and initialize MCPInstrumentation before importing the MCP modules.
const {MCPInstrumentation} = require("@arizeai/openinference-instrumentation-mcp");
const mcpInstrumentation = new MCPInstrumentation();

// Now, manually instrument MCP, as it doesn't have a traditional module structure.
mcpInstrumentation.manuallyInstrument({
    clientModule: require("@modelcontextprotocol/sdk/client/index.js"),
    serverModule: require("@modelcontextprotocol/sdk/server/index.js"),
});

// Finally, configure any other instrumentation and start the sdk
const { getNodeAutoInstrumentations } = require('@opentelemetry/auto-instrumentations-node');
const sdk = new NodeSDK({
    traceExporter: new OTLPTraceExporter(),
    instrumentations: [getNodeAutoInstrumentations()],
});

sdk.start();

process.on("SIGTERM", async () => {
    try {
        await sdk.shutdown();
    } catch (err) {
        console.warn("warning: error shutting down OTel SDK", err);
    }
    process.exit();
});

process.once("beforeExit", async () => {
    // Flush recent telemetry data if about to shutdown.
    try {
        await sdk.shutdown();
    } catch (err) {
        console.warn("warning: error shutting down OTel SDK", err);
    }
});
```

<details>

logs
```
> genai-function-calling@1.0.0 start
> node --env-file .env -r ./telemetry.js index.js --mcp

@opentelemetry/api: Registered a global for diag v1.9.0.
Manually instrumenting @modelcontextprotocol/sdk/client/index
Applying patch for @modelcontextprotocol/sdk/client/index@undefined
Manually instrumenting @modelcontextprotocol/sdk/server/index
Applying patch for @modelcontextprotocol/sdk/server/index@undefined
Loading instrumentation for @opentelemetry/instrumentation-amqplib
Loading instrumentation for @opentelemetry/instrumentation-aws-lambda
@opentelemetry/instrumentation-aws-lambda Skipping lambda instrumentation: no _HANDLER/lambdaHandler or LAMBDA_TASK_ROOT. { taskRoot: undefined, handlerDef: undefined }
Loading instrumentation for @opentelemetry/instrumentation-aws-sdk
Loading instrumentation for @opentelemetry/instrumentation-bunyan
Loading instrumentation for @opentelemetry/instrumentation-cassandra-driver
Loading instrumentation for @opentelemetry/instrumentation-connect
Loading instrumentation for @opentelemetry/instrumentation-cucumber
Loading instrumentation for @opentelemetry/instrumentation-dataloader
Loading instrumentation for @opentelemetry/instrumentation-dns
Loading instrumentation for @opentelemetry/instrumentation-express
Disabling instrumentation for @opentelemetry/instrumentation-fastify
Disabling instrumentation for @opentelemetry/instrumentation-fs
Loading instrumentation for @opentelemetry/instrumentation-generic-pool
Loading instrumentation for @opentelemetry/instrumentation-graphql
Loading instrumentation for @opentelemetry/instrumentation-grpc
Loading instrumentation for @opentelemetry/instrumentation-hapi
Loading instrumentation for @opentelemetry/instrumentation-http
Loading instrumentation for @opentelemetry/instrumentation-ioredis
Loading instrumentation for @opentelemetry/instrumentation-kafkajs
Loading instrumentation for @opentelemetry/instrumentation-knex
Loading instrumentation for @opentelemetry/instrumentation-koa
Loading instrumentation for @opentelemetry/instrumentation-lru-memoizer
Loading instrumentation for @opentelemetry/instrumentation-memcached
Loading instrumentation for @opentelemetry/instrumentation-mongodb
Loading instrumentation for @opentelemetry/instrumentation-mongoose
Loading instrumentation for @opentelemetry/instrumentation-mysql2
Loading instrumentation for @opentelemetry/instrumentation-mysql
Loading instrumentation for @opentelemetry/instrumentation-nestjs-core
Loading instrumentation for @opentelemetry/instrumentation-net
Loading instrumentation for @opentelemetry/instrumentation-pg
Loading instrumentation for @opentelemetry/instrumentation-pino
Loading instrumentation for @opentelemetry/instrumentation-redis
Loading instrumentation for @opentelemetry/instrumentation-redis-4
Loading instrumentation for @opentelemetry/instrumentation-restify
Loading instrumentation for @opentelemetry/instrumentation-router
Loading instrumentation for @opentelemetry/instrumentation-runtime-node
Loading instrumentation for @opentelemetry/instrumentation-socket.io
Loading instrumentation for @opentelemetry/instrumentation-tedious
Loading instrumentation for @opentelemetry/instrumentation-undici
Loading instrumentation for @opentelemetry/instrumentation-winston
Current logger will be overwritten from Error
    at DiagAPI.setLogger (/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/node_modules/@opentelemetry/api/build/src/api/diag.js:65:37)
    at new NodeSDK (/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/node_modules/@opentelemetry/sdk-node/build/src/sdk.js:130:24)
    at Object.<anonymous> (/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/telemetry.js:20:13)
    at Module._compile (node:internal/modules/cjs/loader:1554:14)
    at Object..js (node:internal/modules/cjs/loader:1706:10)
    at Module.load (node:internal/modules/cjs/loader:1289:32)
    at Function._load (node:internal/modules/cjs/loader:1108:12)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:220:24)
    at Module.<anonymous> (node:internal/modules/cjs/loader:1311:12)
Current logger will overwrite one already registered from Error
    at DiagAPI.setLogger (/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/node_modules/@opentelemetry/api/build/src/api/diag.js:65:37)
    at new NodeSDK (/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/node_modules/@opentelemetry/sdk-node/build/src/sdk.js:130:24)
    at Object.<anonymous> (/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/telemetry.js:20:13)
    at Module._compile (node:internal/modules/cjs/loader:1554:14)
    at Object..js (node:internal/modules/cjs/loader:1706:10)
    at Module.load (node:internal/modules/cjs/loader:1289:32)
    at Function._load (node:internal/modules/cjs/loader:1108:12)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:220:24)
    at Module.<anonymous> (node:internal/modules/cjs/loader:1311:12)
@opentelemetry/api: Registered a global for diag v1.9.0.
OTEL_LOGS_EXPORTER is empty. Using default otlp exporter.
@opentelemetry/api: Registered a global for trace v1.9.0.
@opentelemetry/api: Registered a global for context v1.9.0.
@opentelemetry/api: Registered a global for propagation v1.9.0.
EnvDetectorSync found resource. Resource {
  _attributes: { 'service.name': 'genai-function-calling' },
  asyncAttributesPending: false,
  _syncAttributes: { 'service.name': 'genai-function-calling' },
  _asyncAttributesPromise: undefined
}
OSDetectorSync found resource. Resource {
  _attributes: { 'os.type': 'darwin', 'os.version': '24.4.0' },
  asyncAttributesPending: false,
  _syncAttributes: { 'os.type': 'darwin', 'os.version': '24.4.0' },
  _asyncAttributesPromise: undefined
}
ServiceInstanceIdDetectorSync found resource. Resource {
  _attributes: { 'service.instance.id': 'a30bb5e3-4470-405a-ac4d-34103746233f' },
  asyncAttributesPending: false,
  _syncAttributes: { 'service.instance.id': 'a30bb5e3-4470-405a-ac4d-34103746233f' },
  _asyncAttributesPromise: undefined
}
ProcessDetectorSync found resource. Resource {
  _attributes: {
    'process.pid': 72997,
    'process.executable.name': 'node',
    'process.executable.path': '/Users/adriancole/.nvm/versions/node/v22.14.0/bin/node',
    'process.command_args': [
      '/Users/adriancole/.nvm/versions/node/v22.14.0/bin/node',
      '--env-file',
      '.env',
      '-r',
      './telemetry.js',
      '/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/index.js',
      '--mcp'
    ],
    'process.runtime.version': '22.14.0',
    'process.runtime.name': 'nodejs',
    'process.runtime.description': 'Node.js',
    'process.command': '/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/index.js',
    'process.owner': 'adriancole'
  },
  asyncAttributesPending: false,
  _syncAttributes: {
    'process.pid': 72997,
    'process.executable.name': 'node',
    'process.executable.path': '/Users/adriancole/.nvm/versions/node/v22.14.0/bin/node',
    'process.command_args': [
      '/Users/adriancole/.nvm/versions/node/v22.14.0/bin/node',
      '--env-file',
      '.env',
      '-r',
      './telemetry.js',
      '/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/index.js',
      '--mcp'
    ],
    'process.runtime.version': '22.14.0',
    'process.runtime.name': 'nodejs',
    'process.runtime.description': 'Node.js',
    'process.command': '/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/index.js',
    'process.owner': 'adriancole'
  },
  _asyncAttributesPromise: undefined
}
HostDetectorSync found resource. Resource {
  _attributes: {
    'host.name': 'Adrians-MacBook-Pro.local',
    'host.arch': 'arm64',
    'host.id': '0BD5E4F8-C9FD-5300-9F52-36C38F9D1C54'
  },
  asyncAttributesPending: false,
  _syncAttributes: { 'host.name': 'Adrians-MacBook-Pro.local', 'host.arch': 'arm64' },
  _asyncAttributesPromise: Promise { { 'host.id': '0BD5E4F8-C9FD-5300-9F52-36C38F9D1C54' } }
}
Current logger will be overwritten from Error
    at DiagAPI.setLogger (/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/node_modules/@opentelemetry/api/build/src/api/diag.js:65:37)
    at new NodeSDK (/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/node_modules/@opentelemetry/sdk-node/build/src/sdk.js:130:24)
    at Object.<anonymous> (/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/telemetry.js:20:13)
    at Module._compile (node:internal/modules/cjs/loader:1554:14)
    at Object..js (node:internal/modules/cjs/loader:1706:10)
    at Module.load (node:internal/modules/cjs/loader:1289:32)
    at Function._load (node:internal/modules/cjs/loader:1108:12)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:220:24)
    at Module.<anonymous> (node:internal/modules/cjs/loader:1311:12)
Current logger will overwrite one already registered from Error
    at DiagAPI.setLogger (/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/node_modules/@opentelemetry/api/build/src/api/diag.js:65:37)
    at new NodeSDK (/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/node_modules/@opentelemetry/sdk-node/build/src/sdk.js:130:24)
    at Object.<anonymous> (/Users/adriancole/oss/observability-examples/genai-function-calling/vercel-ai/telemetry.js:20:13)
    at Module._compile (node:internal/modules/cjs/loader:1554:14)
    at Object..js (node:internal/modules/cjs/loader:1706:10)
    at Module.load (node:internal/modules/cjs/loader:1289:32)
    at Function._load (node:internal/modules/cjs/loader:1108:12)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:220:24)
    at Module.<anonymous> (node:internal/modules/cjs/loader:1311:12)
The latest version of Elasticsearch 8 is 8.18.0.
OTLPExportDelegate items to be sent [
  Span {
    attributes: {
      'http.request.method': 'POST',
      'http.request.method_original': 'POST',
      'url.full': 'http://localhost:11434/v1/chat/completions',
      'url.path': '/v1/chat/completions',
      'url.query': '',
      'url.scheme': 'http',
      'server.address': 'localhost',
      'server.port': 11434,
      'user_agent.original': 'node',
      'network.peer.address': '127.0.0.1',
      'network.peer.port': 11434,
      'http.response.status_code': 200,
      'http.response.header.content-length': 446
    },
    links: [],
    events: [],
    _droppedAttributesCount: 0,
    _droppedEventsCount: 0,
    _droppedLinksCount: 0,
    status: { code: 0 },
    endTime: [ 1745371658, 860247667 ],
    _ended: true,
    _duration: [ 1, 21247667 ],
    name: 'POST',
    _spanContext: {
      traceId: '1470598e56da6475cfb9691ae25e22f6',
      spanId: '64a889227ea9bb9e',
      traceFlags: 1,
      traceState: undefined
    },
    parentSpanId: '20cd461ad528be2f',
    kind: 2,
    _performanceStartTime: 1440.817958,
    _performanceOffset: -0.950927734375,
    _startTimeProvided: false,
    startTime: [ 1745371657, 839000000 ],
    resource: Resource {
      _attributes: [Object],
      asyncAttributesPending: false,
      _syncAttributes: [Object],
      _asyncAttributesPromise: [Promise]
    },
    instrumentationLibrary: {
      name: '@opentelemetry/instrumentation-undici',
      version: '0.11.0',
      schemaUrl: undefined
    },
    _spanLimits: {
      attributeValueLengthLimit: Infinity,
      attributeCountLimit: 128,
      linkCountLimit: 128,
      eventCountLimit: 128,
      attributePerEventCountLimit: 128,
      attributePerLinkCountLimit: 128
    },
    _attributeValueLengthLimit: Infinity,
    _spanProcessor: MultiSpanProcessor { _spanProcessors: [Array] }
  },
  Span {
    attributes: {
      'operation.name': 'ai.generateText.doGenerate',
      'ai.operationId': 'ai.generateText.doGenerate',
      'ai.model.provider': 'openai.chat',
      'ai.model.id': 'qwen2.5',
      'ai.settings.temperature': 0,
      'ai.settings.tool_choice': 'auto',
      'ai.settings.maxRetries': 2,
      'ai.prompt.format': 'messages',
      'ai.prompt.messages': '[{"role":"user","content":[{"type":"text","text":"What is the latest version of Elasticsearch 8?"}]}]',
      'ai.prompt.tools': [Array],
      'ai.prompt.toolChoice': '{"type":"auto"}',
      'gen_ai.system': 'openai.chat',
      'gen_ai.request.model': 'qwen2.5',
      'gen_ai.request.temperature': 0,
      'ai.response.finishReason': 'tool-calls',
      'ai.response.text': '',
      'ai.response.toolCalls': '[{"toolCallType":"function","toolCallId":"call_8k9le99q","toolName":"get_latest_elasticsearch_version","args":"{\\"majorVersion\\":8}"}]',
      'ai.response.id': 'chatcmpl-459',
      'ai.response.model': 'qwen2.5',
      'ai.response.timestamp': '2025-04-23T01:27:38.000Z',
      'ai.usage.promptTokens': 185,
      'ai.usage.completionTokens': 24,
      'gen_ai.response.finish_reasons': [Array],
      'gen_ai.response.id': 'chatcmpl-459',
      'gen_ai.response.model': 'qwen2.5',
      'gen_ai.usage.input_tokens': 185,
      'gen_ai.usage.output_tokens': 24
    },
    links: [],
    events: [],
    _droppedAttributesCount: 0,
    _droppedEventsCount: 0,
    _droppedLinksCount: 0,
    status: { code: 0 },
    endTime: [ 1745371658, 870031958 ],
    _ended: true,
    _duration: [ 1, 39031958 ],
    name: 'ai.generateText.doGenerate',
    _spanContext: {
      traceId: '1470598e56da6475cfb9691ae25e22f6',
      spanId: '20cd461ad528be2f',
      traceFlags: 1,
      traceState: undefined
    },
    parentSpanId: 'ab6bec3a78044b10',
    kind: 0,
    _performanceStartTime: 1432.268375,
    _performanceOffset: -0.4013671875,
    _startTimeProvided: false,
    startTime: [ 1745371657, 831000000 ],
    resource: Resource {
      _attributes: [Object],
      asyncAttributesPending: false,
      _syncAttributes: [Object],
      _asyncAttributesPromise: [Promise]
    },
    instrumentationLibrary: { name: 'ai', version: undefined, schemaUrl: undefined },
    _spanLimits: {
      attributeValueLengthLimit: Infinity,
      attributeCountLimit: 128,
      linkCountLimit: 128,
      eventCountLimit: 128,
      attributePerEventCountLimit: 128,
      attributePerLinkCountLimit: 128
    },
    _attributeValueLengthLimit: Infinity,
    _spanProcessor: MultiSpanProcessor { _spanProcessors: [Array] }
  },
  Span {
    attributes: {
      'operation.name': 'ai.toolCall',
      'ai.operationId': 'ai.toolCall',
      'ai.toolCall.name': 'get_latest_elasticsearch_version',
      'ai.toolCall.id': 'call_8k9le99q',
      'ai.toolCall.args': '{"majorVersion":8}',
      'ai.toolCall.result': '{"content":[{"type":"text","text":"8.18.0"}]}'
    },
    links: [],
    events: [],
    _droppedAttributesCount: 0,
    _droppedEventsCount: 0,
    _droppedLinksCount: 0,
    status: { code: 0 },
    endTime: [ 1745371659, 56501250 ],
    _ended: true,
    _duration: [ 0, 185501250 ],
    name: 'ai.toolCall',
    _spanContext: {
      traceId: '1470598e56da6475cfb9691ae25e22f6',
      spanId: 'c0d2bfb535d95117',
      traceFlags: 1,
      traceState: undefined
    },
    parentSpanId: 'ab6bec3a78044b10',
    kind: 0,
    _performanceStartTime: 2472.615375,
    _performanceOffset: -0.74853515625,
    _startTimeProvided: false,
    startTime: [ 1745371658, 871000000 ],
    resource: Resource {
      _attributes: [Object],
      asyncAttributesPending: false,
      _syncAttributes: [Object],
      _asyncAttributesPromise: [Promise]
    },
    instrumentationLibrary: { name: 'ai', version: undefined, schemaUrl: undefined },
    _spanLimits: {
      attributeValueLengthLimit: Infinity,
      attributeCountLimit: 128,
      linkCountLimit: 128,
      eventCountLimit: 128,
      attributePerEventCountLimit: 128,
      attributePerLinkCountLimit: 128
    },
    _attributeValueLengthLimit: Infinity,
    _spanProcessor: MultiSpanProcessor { _spanProcessors: [Array] }
  },
  Span {
    attributes: {
      'http.request.method': 'POST',
      'http.request.method_original': 'POST',
      'url.full': 'http://localhost:11434/v1/chat/completions',
      'url.path': '/v1/chat/completions',
      'url.query': '',
      'url.scheme': 'http',
      'server.address': 'localhost',
      'server.port': 11434,
      'user_agent.original': 'node',
      'network.peer.address': '127.0.0.1',
      'network.peer.port': 11434,
      'http.response.status_code': 200,
      'http.response.header.content-length': 332
    },
    links: [],
    events: [],
    _droppedAttributesCount: 0,
    _droppedEventsCount: 0,
    _droppedLinksCount: 0,
    status: { code: 0 },
    endTime: [ 1745371659, 905519792 ],
    _ended: true,
    _duration: [ 0, 847519792 ],
    name: 'POST',
    _spanContext: {
      traceId: '1470598e56da6475cfb9691ae25e22f6',
      spanId: '5f898f78963862d3',
      traceFlags: 1,
      traceState: undefined
    },
    parentSpanId: '531a50e0b0a3e65c',
    kind: 2,
    _performanceStartTime: 2659.513708,
    _performanceOffset: -0.646728515625,
    _startTimeProvided: false,
    startTime: [ 1745371659, 58000000 ],
    resource: Resource {
      _attributes: [Object],
      asyncAttributesPending: false,
      _syncAttributes: [Object],
      _asyncAttributesPromise: [Promise]
    },
    instrumentationLibrary: {
      name: '@opentelemetry/instrumentation-undici',
      version: '0.11.0',
      schemaUrl: undefined
    },
    _spanLimits: {
      attributeValueLengthLimit: Infinity,
      attributeCountLimit: 128,
      linkCountLimit: 128,
      eventCountLimit: 128,
      attributePerEventCountLimit: 128,
      attributePerLinkCountLimit: 128
    },
    _attributeValueLengthLimit: Infinity,
    _spanProcessor: MultiSpanProcessor { _spanProcessors: [Array] }
  },
  Span {
    attributes: {
      'operation.name': 'ai.generateText.doGenerate',
      'ai.operationId': 'ai.generateText.doGenerate',
      'ai.model.provider': 'openai.chat',
      'ai.model.id': 'qwen2.5',
      'ai.settings.temperature': 0,
      'ai.settings.tool_choice': 'auto',
      'ai.settings.maxRetries': 2,
      'ai.prompt.format': 'messages',
      'ai.prompt.messages': '[{"role":"user","content":[{"type":"text","text":"What is the latest version of Elasticsearch 8?"}]},{"role":"assistant","content":[{"type":"tool-call","toolCallId":"call_8k9le99q","toolName":"get_latest_elasticsearch_version","args":{"majorVersion":8}}]},{"role":"tool","content":[{"type":"tool-result","toolCallId":"call_8k9le99q","toolName":"get_latest_elasticsearch_version","result":{"content":[{"type":"text","text":"8.18.0"}]}}]}]',
      'ai.prompt.tools': [Array],
      'ai.prompt.toolChoice': '{"type":"auto"}',
      'gen_ai.system': 'openai.chat',
      'gen_ai.request.model': 'qwen2.5',
      'gen_ai.request.temperature': 0,
      'ai.response.finishReason': 'stop',
      'ai.response.text': 'The latest version of Elasticsearch 8 is 8.18.0.',
      'ai.response.id': 'chatcmpl-988',
      'ai.response.model': 'qwen2.5',
      'ai.response.timestamp': '2025-04-23T01:27:39.000Z',
      'ai.usage.promptTokens': 220,
      'ai.usage.completionTokens': 17,
      'gen_ai.response.finish_reasons': [Array],
      'gen_ai.response.id': 'chatcmpl-988',
      'gen_ai.response.model': 'qwen2.5',
      'gen_ai.usage.input_tokens': 220,
      'gen_ai.usage.output_tokens': 17
    },
    links: [],
    events: [],
    _droppedAttributesCount: 0,
    _droppedEventsCount: 0,
    _droppedLinksCount: 0,
    status: { code: 0 },
    endTime: [ 1745371659, 906256042 ],
    _ended: true,
    _duration: [ 0, 849256042 ],
    name: 'ai.generateText.doGenerate',
    _spanContext: {
      traceId: '1470598e56da6475cfb9691ae25e22f6',
      spanId: '531a50e0b0a3e65c',
      traceFlags: 1,
      traceState: undefined
    },
    parentSpanId: 'ab6bec3a78044b10',
    kind: 0,
    _performanceStartTime: 2658.837708,
    _performanceOffset: -0.970703125,
    _startTimeProvided: false,
    startTime: [ 1745371659, 57000000 ],
    resource: Resource {
      _attributes: [Object],
      asyncAttributesPending: false,
      _syncAttributes: [Object],
      _asyncAttributesPromise: [Promise]
    },
    instrumentationLibrary: { name: 'ai', version: undefined, schemaUrl: undefined },
    _spanLimits: {
      attributeValueLengthLimit: Infinity,
      attributeCountLimit: 128,
      linkCountLimit: 128,
      eventCountLimit: 128,
      attributePerEventCountLimit: 128,
      attributePerLinkCountLimit: 128
    },
    _attributeValueLengthLimit: Infinity,
    _spanProcessor: MultiSpanProcessor { _spanProcessors: [Array] }
  },
  Span {
    attributes: {
      'operation.name': 'ai.generateText',
      'ai.operationId': 'ai.generateText',
      'ai.model.provider': 'openai.chat',
      'ai.model.id': 'qwen2.5',
      'ai.settings.temperature': 0,
      'ai.settings.tool_choice': 'auto',
      'ai.settings.maxRetries': 2,
      'ai.prompt': '{"messages":[{"role":"user","content":"What is the latest version of Elasticsearch 8?"}]}',
      'ai.settings.maxSteps': 10,
      'ai.response.finishReason': 'stop',
      'ai.response.text': 'The latest version of Elasticsearch 8 is 8.18.0.',
      'ai.usage.promptTokens': 220,
      'ai.usage.completionTokens': 17
    },
    links: [],
    events: [],
    _droppedAttributesCount: 0,
    _droppedEventsCount: 0,
    _droppedLinksCount: 0,
    status: { code: 0 },
    endTime: [ 1745371659, 907026042 ],
    _ended: true,
    _duration: [ 2, 77026042 ],
    name: 'ai.generateText',
    _spanContext: {
      traceId: '1470598e56da6475cfb9691ae25e22f6',
      spanId: 'ab6bec3a78044b10',
      traceFlags: 1,
      traceState: undefined
    },
    parentSpanId: undefined,
    kind: 0,
    _performanceStartTime: 1431.324333,
    _performanceOffset: -0.457275390625,
    _startTimeProvided: false,
    startTime: [ 1745371657, 830000000 ],
    resource: Resource {
      _attributes: [Object],
      asyncAttributesPending: false,
      _syncAttributes: [Object],
      _asyncAttributesPromise: [Promise]
    },
    instrumentationLibrary: { name: 'ai', version: undefined, schemaUrl: undefined },
    _spanLimits: {
      attributeValueLengthLimit: Infinity,
      attributeCountLimit: 128,
      linkCountLimit: 128,
      eventCountLimit: 128,
      attributePerEventCountLimit: 128,
      attributePerLinkCountLimit: 128
    },
    _attributeValueLengthLimit: Infinity,
    _spanProcessor: MultiSpanProcessor { _spanProcessors: [Array] }
  }
]
@opentelemetry/instrumentation-http Applying instrumentation patch for nodejs core module on require hook { module: 'http' }
@opentelemetry/instrumentation-http Applying instrumentation patch for nodejs core module on require hook { module: 'https' }
Instrumentation suppressed, returning Noop Span
@opentelemetry/instrumentation-http http instrumentation outgoingRequest
@opentelemetry/instrumentation-http http.ClientRequest return request
OTLPExportDelegate shutdown started
@opentelemetry/instrumentation-http outgoingRequest on response()
@opentelemetry/instrumentation-http outgoingRequest on end()
@opentelemetry/instrumentation-http outgoingRequest on request close()
OTLPExportDelegate shutdown started
```
</details>

codefromthecrypt commented:
for this I'll use docker for now on (`docker compose run --build --rm genai-function-calling -- --mcp`). all commits have split traces from my view anyway. I'm using otel-tui version 0.4.10 and docker controls npm etc at this point.

<details>

ec03ebf222dbeac73772aa8f9785cf455fa1521c 
![Screenshot 2025-04-23 at 9 52 36 AM](https://github.com/user-attachments/assets/62befeab-e345-46dd-9dc6-796ebcde4166)


3de0d7038f55bdd79f9b7533389aac463d50cac3  ![Screenshot 2025-04-23 at 9 49 01 AM](https://github.com/user-attachments/assets/199456bd-e353-4377-8655-6fae046d9fc8)

a143b7a0d7e75b9abfdcda7d46d60f62e5dbd2fa ![Screenshot 2025-04-23 at 9 54 37 AM](https://github.com/user-attachments/assets/86cdeb0b-82d6-4fe4-89c1-5b6308196722)

a5719d725bdd31eec4427fe74234691beee72ddc ![Screenshot 2025-04-23 at 9 57 22 AM](https://github.com/user-attachments/assets/af1920ef-1873-488f-8ada-944268e14ba1)

</details>

codefromthecrypt commented:
update: seems to be more about the path vercel is using mcp being different than tested paths. @anuraaga will chase it up tomorrow

codefromthecrypt commented:
ok all good with the tiniest diff!
